src/java.base/share/classes/java/lang/invoke/MethodHandleImpl.java

```diff
@@ -2089,21 +1876,64 @@ static Object tryFinally(MethodHandle target, MethodHandle cleanup, Object... av
         return r;
     }
 
+    // see varargsArray method for chaching/package-private version of this
+    private static MethodHandle makeCollector(Class<?> arrayType, int parameterCount) {
+        MethodType type = MethodType.methodType(arrayType, Collections.nCopies(parameterCount, arrayType.componentType()));
+
+        MethodHandle storeFunc = makeArrayElementAccessor(arrayType, ArrayAccess.SET);
+        LambdaForm form = makeCollectorForm(type.basicType(), arrayType.componentType(), storeFunc);
+
+        return SimpleMethodHandle.make(type, form);
+    }
+
+    private static LambdaForm makeCollectorForm(MethodType basicType, Class<?> componentType, MethodHandle storeFunc) {
+        MethodType lambdaType = basicType.invokerType();
+        int parameterCount = basicType.parameterCount();
+
+        // It might be possible to share this lambda form for all reference types - since they
+        // use the same underlying store function - by injecting an array creation handle as a
+        // field of the receiver handle.
+        // This is left as a followup enhancement, as it needs to be investigated if this causes
+        // profile pollution.
```

Comment:
Profile pollution shouldn't be a problem here: both types (element type + array type) will be constants attached to the "receiver" BMH instance and during inlining will be fetched from there.  

I'm fine with handling it as a separate RFE.

Reply:
That's what I thought as well (but not 100% sure). I can partially roll back the last commit so the code still uses an injected array constructor handle, and then it should be easy to add caching in the cases where the component type is a reference type.

Reply:
Whatever you prefer. I'm fine with it either way.

Reply:
I found it easier to add sharing now, since it makes the intrinsic redundant (and it can be removed again).